src/protocol/sort/shuffle.rs

```diff
@@ -228,88 +213,187 @@ mod tests {
+        };
+        use crate::protocol::QueryId;
+        use crate::test_fixture::{Reconstruct, Runner, TestWorld};
+
+        #[tokio::test]
+        async fn semi_honest() {
+            const BATCHSIZE: u8 = 25;
+            let world = TestWorld::new(QueryId);
+
+            let input: Vec<u8> = (0..BATCHSIZE).collect();
+            let hashed_input: HashSet<u8> = input.clone().into_iter().collect();
+
+            let input_u128: Vec<u128> = input.iter().map(|x| u128::from(*x)).collect();
+
+            let result = world
+                .semi_honest(
+                    input_u128.clone().into_iter().map(Fp31::from),
+                    |ctx, m_shares| async move {
+                        let perms =
+                            get_two_of_three_random_permutations(BATCHSIZE.into(), ctx.prss_rng());
+                        shuffle_shares(
+                            m_shares,
+                            (perms.0.as_slice(), perms.1.as_slice()),
+                            ctx.clone(),
+                        )
+                        .await
+                        .unwrap()
+                    },
+                )
+                .await;
+
+            let mut hashed_output_secret = HashSet::new();
+            let mut output_secret = Vec::new();
+            for (r0, (r1, r2)) in zip(result[0].iter(), zip(result[1].iter(), result[2].iter())) {
+                let val = (r0, r1, r2).reconstruct();
+                output_secret.push(u8::from(val));
+                hashed_output_secret.insert(u8::from(val));
+            }
+
+            // Secrets should be shuffled
+            assert_ne!(output_secret, input);
+
+            // Shuffled output should have same inputs
+            assert_eq!(hashed_output_secret, hashed_input);
         }
 
-        // Order of shares should now be different from original
-        assert_ne!(results, original);
-        // Secrets should be shuffled also
-        assert_ne!(output_secret, input);
-
-        // Shuffled output should have same inputs
-        assert_eq!(hashed_output_secret, hashed_input);
+        #[tokio::test]
+        async fn shuffle_unshuffle() {
+            const BATCHSIZE: usize = 5;
+
+            let world = TestWorld::new(QueryId);
+            let input: Vec<u128> = (0..u128::try_from(BATCHSIZE).unwrap()).collect();
+
+            let result = world
+                .semi_honest(
+                    input.clone().into_iter().map(Fp31::from),
+                    |ctx, m_shares| async move {
+                        let perms = get_two_of_three_random_permutations(
+                            BATCHSIZE.try_into().unwrap(),
+                            ctx.prss_rng(),
+                        );
+                        let shuffled = shuffle_shares(
+                            m_shares,
+                            (perms.0.as_slice(), perms.1.as_slice()),
+                            ctx.clone(),
```

Comment:
If you narrow for the first one, you can just use the context for the second without cloning.

Reply:
Except that I now see that you are getting the PRSS from this context already...